nx/lib/nx/binary_backend/matrix.ex

```diff
@@ -377,16 +381,19 @@ defmodule Nx.BinaryBackend.Matrix do
             if is_nil(q) do
               h
             else
-              dot_matrix(q, h)
+              multiply_matrix(q, h)
```

Comment:
Is the `h` matrix complex? wouldn't taking the real part of that suffice?
using `multiply_matrix` with 2 complex matrices can lead to problems

Reply:
`q`, `h` may also have complex numbers as elements. What problems may arise?

Reply:
The QR definition is wrong, for instance.

The algorithm is defined such that each iteration builds Q and R by composing H_i matrices over a pair of matrices, such that premultiplying with H_i zeroes out the column i under the diagonal.

So we have {I, A} -> {H1, H1*.A} -> {H1.H2, H2*.H1*.A} and so on until {Q, R}. But the definition is only correct if Q = H1.H2...Hn and R = Hn*....H2*.H1.A, because when doing Q.R you get the H_i.H_i* = I pairs to cancel out. That is:

Q.R = H1.H2...(Hn.Hn*)....H2*.H1.A = H1.(H2.H2*).H1*.A = H1.H1*.A = A

By using "multiply_matrix" instead of "dot_matrix", you're removing that property.
If something went wrong, it's probably on line 382 because there's an implicit dot_matrix(identity, h) there, that is, you must return adjoint(h) there

Reply:
same implicit dot_matrix at line 220

Reply:
I changed the function in lines 220 and 384 to `dot_matrix` as [the formula shown](https://github.com/elixir-nx/nx/pull/933#discussion_r982887087). However, an error occurred in tests such as `eigh` using the function of QR decomposition. On the other hand, if the functions in lines 220 and 384 are left as `dot_matrix_real`, no error occurs.

Reply:
I checked your branch out locally, and there's for sure some inconsistency with the householder reflector that turns that dot_matrix_real necessary, but we can keep it for now.

Reply:
Thanks for the advice. Yes, I agree.
I added the TODO comments and discussion URL so we don't forget.  [5e670a7](https://github.com/elixir-nx/nx/pull/933/commits/5e670a7d91485623ebcf9407f01d43d4b7ef2ce7)